### PR TITLE
Add App.js and top-level react-router

### PR DESCRIPTION
**Pull request by kevinrobinson**
# Who is this PR for?
developers

# What problem does this PR fix?
Answers "How do I add a new page?" and creates a migration path for removing code in `legacyRouteHandlers`.  It came out of building towards https://github.com/studentinsights/studentinsights/issues/1497.

# What does this PR do?
Changes the top-level `index.js` file so that it checks for a `main` element, runs through `legacyRouteHandler`, and then if not it mounts the new `App.js` on `main`.  This also adds a placeholder route for `/home`, which is for the new home page prototype.

On the server side, also adds a controller action for the server to respond with on client-side routes, `ui#ui`, which sends back just the current educator and then leaves any other data fetching up to the client.  So for example, on a request to `/home`, Rails sends down the same application layout with the nav bar and embeds the `current_educator` data in `serialized-data`.  The JS code then runs, tries out the legacy routes (none match), then mounts `App.js` which reads the minimal generic `serializedData` and then renders `HomePage`.  It's up to `HomePage` to request what other data it needs.

# Screenshot (if adding a client-side feature)
(note that this is placeholder; this PR is about the routing mechanism)
screen shot 2018-03-02 at 12 12 18 pm

# Checklists
## Javascript QA
+ [x] Author checked latest in IE - Student Profile
+ [x] Author checked latest in IE - Home page
+ [x] Author checked latest in IE - Import records
+ [x] Author checked latest in IE - Bulk upload
+ [x] Author checked latest in IE - School Overview
+ [x] Author checked latest in IE - Section page
+ [x] Author checked latest in IE - Homeroom page

## Automated Testing
+ [x] Author included specs for new code

### CODE REVIEW

**Comment by kevinrobinson** [comment]
The idea here is that this could be how we add new client-side routes for new pages, so I asked everybody to review and make sure this seems reasonable, and to help surface any questions or other ideas about this too.  Thanks y'all!

File: ui/index.js [hunk 1]
```diff
@@ -25,4 +29,12 @@ if ($('.student-searchbar').length > 0) {
 }
 
 // Routing
-route();
+// Some pages are server-rendered and have a different structure
+// other than #main so we ignore those.  Other pages add in class names
+// to the body tag that `legacyRouteHandler` works with.  Newer pages
+// should handle routing with react-router inside the `App` component.
+const mainEl = document.getElementById('main');
+if (mainEl) {
```
Comment (alexsoble) [inline ui/index.js]: Is `index.js` going to run on every single page/page load in the app?

Comment (kevinrobinson) [reply ui/index.js]: Yep, it already does except for a few admin pages that don't load any JS at all.

Ideally we'd run the routing, then run any other code based on that.  But we're still doing a lot of JS execution in the different legacy.js files as part of the migration from Sprockets.

**Comment by alexsoble** [comment]
@kevinrobinson I want to make a flow chart to show how routing will work with this PR! I've read through the code a few times and the core concepts make sense, I think a visual will help clarify the flow.

**Comment by kevinrobinson** [comment]
@alexsoble @edavidsonsawyer @gabivoicu I'm planning on shipping this today, so if you have other comments or feedback please feel free to share them either before or after this is merged.

Reviewing is a good place to ask questions, learn about stuff, and clarify what we're all building together over time, and I'm a big fan of after-the-fact reviews too especially when people are getting other stuff shipped too :)

**Comment by alexsoble** [comment]
@kevinrobinson and I took some time to walk through the JS, diagram of routing coming soon! 

In the meantime, 🚢 🚢 🚢! 😎

**Pull request merged**